pulsar/consumer_test.go

```diff
@@ -0,0 +1,414 @@
+
+	consumer1, err := client.Subscribe(ConsumerOptions{
+		Topic:            topic,
+		SubscriptionName: "sub-1",
+		Type:             KeyShared,
+	})
+	assert.Nil(t, err)
+	defer consumer1.Close()
+
+	consumer2, err := client.Subscribe(ConsumerOptions{
+		Topic:            topic,
+		SubscriptionName: "sub-1",
+		Type:             KeyShared,
+	})
+	assert.Nil(t, err)
+	defer consumer2.Close()
+
+	// create producer
+	producer, err := client.CreateProducer(ProducerOptions{
+		Topic: topic,
+	})
+	assert.Nil(t, err)
+	defer producer.Close()
+
+	ctx := context.Background()
+	for i := 0; i < 10; i++ {
+		err := producer.Send(ctx, &ProducerMessage{
+			Key:     fmt.Sprintf("key-shared-%d", i%3),
+			Payload: []byte(fmt.Sprintf("value-%d", i)),
+		})
+		assert.Nil(t, err)
+	}
+
+	time.Sleep(time.Second * 5)
+
+	go func() {
+		for i := 0; i < 10; i++ {
+			msg, err := consumer1.Receive(ctx)
+			assert.Nil(t, err)
+			if msg != nil {
+				fmt.Printf("consumer1 key is: %s, value is: %s\n", msg.Key(), string(msg.Payload()))
+				err = consumer1.Ack(msg)
+				assert.Nil(t, err)
+			}
+		}
+	}()
+
+	go func() {
+		for i := 0; i < 10; i++ {
+			msg2, err := consumer2.Receive(ctx)
+			assert.Nil(t, err)
+			if msg2 != nil {
+				fmt.Printf("consumer2 key is:%s, value is: %s\n", msg2.Key(), string(msg2.Payload()))
+				err = consumer2.Ack(msg2)
+				assert.Nil(t, err)
+			}
+		}
+	}()
+}
+
+func TestPartitionTopicsConsumerPubSub(t *testing.T) {
+	client, err := NewClient(ClientOptions{
+		URL: lookupURL,
+	})
+	assert.Nil(t, err)
+	defer client.Close()
+
+	topic := "persistent://public/default/testGetPartitions"
+	testURL := adminURL + "/" + "admin/v2/persistent/public/default/testGetPartitions/partitions"
+
+	makeHTTPCall(t, http.MethodPut, testURL, "3")
+
+	// create producer
+	producer, err := client.CreateProducer(ProducerOptions{
+		Topic: topic,
+	})
+	assert.Nil(t, err)
+	defer producer.Close()
+
+	topics, err := client.TopicPartitions(topic)
+	assert.Nil(t, err)
+	assert.Equal(t, topic+"-partition-0", topics[0])
+	assert.Equal(t, topic+"-partition-1", topics[1])
+	assert.Equal(t, topic+"-partition-2", topics[2])
+
+	consumer, err := client.Subscribe(ConsumerOptions{
+		Topic:            topic,
+		SubscriptionName: "my-sub",
+		Type:             Exclusive,
+	})
+	assert.Nil(t, err)
+	defer consumer.Close()
+
+	ctx := context.Background()
+	for i := 0; i < 10; i++ {
+		err := producer.Send(ctx, &ProducerMessage{
+			Payload: []byte(fmt.Sprintf("hello-%d", i)),
+		})
+		assert.Nil(t, err)
+	}
+
+	msgs := make([]string, 0)
+
+	for i := 0; i < 10; i++ {
+		msg, err := consumer.Receive(ctx)
+		assert.Nil(t, err)
+		msgs = append(msgs, string(msg.Payload()))
+
+		fmt.Printf("Received message msgId: %#v -- content: '%s'\n",
+			msg.ID(), string(msg.Payload()))
+
+		if err := consumer.Ack(msg); err != nil {
+			assert.Nil(t, err)
+		}
+	}
+
+	assert.Equal(t, len(msgs), 10)
+}
+
+func TestConsumerAckTimeout(t *testing.T) {
```

Comment:
I would suggest add tests that ack partially of the messages, and only non-acked messages will be redeliver.
Also add this kind of test for shared multi-partitions consumer.

Reply:
The issue will fix in https://github.com/apache/pulsar-client-go/issues/35